File: 50.pow-x-n.java
/*
 * @lc app=leetcode.cn id=50 lang=java
 *
 * [50] Pow(x, n)
 */

// @lc code=start
class Solution {
    public double myPow(double x, int n) {
        long N = n;
        if (N < 0) {
            N = -N;
            x = 1 / x;
        }
        double ans = 1;
        double current = x;
        while (N > 0) {
            if ((N & 1) == 1) {
                ans *= current;
            }
            current = current * current;
            N /= 2;
        }
        return ans;
    }
}
// @lc code=end

File: prepare/50-powx-n.go
package main

func myPow(x float64, n int) float64 {
	if n < 0 {
		n = -n
		x = 1 / x
	}
	cur, ans := x, float64(1)
	for n > 0 {
		if (n & 1) == 1 {
			ans *= cur
		}
		cur *= cur
		n /= 2
	}
	return ans
}
